invenio_rdm_records/services/components/internal_notes.py

```diff
@@ -0,0 +1,64 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2024 CERN.
+#
+# Invenio-RDM-Records is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+
+"""RDM service component for custom fields."""
+from copy import deepcopy, copy
+from datetime import datetime, timezone
+
+from invenio_i18n import lazy_gettext as _
+from invenio_pidstore.providers.recordid_v2 import RecordIdProviderV2
+from marshmallow import ValidationError
+
+from .metadata import MetadataComponent
+
+
+class InternalNotesComponent(MetadataComponent):
+    """Service component for custom fields."""
+
+    field = "internal_notes"
+    new_version_skip_fields = []
+
+    def create(self, identity, data=None, record=None, **kwargs):
+        """Inject note to the record."""
+        notes = data.get(self.field, [])
+        for note in notes:
+            note.setdefault("added_by", {"user": identity.id})
+            note.setdefault("timestamp", datetime.now(timezone.utc).isoformat())
+        record.update({"internal_notes": notes})
+
+    def update_draft(self, identity, data=None, record=None, **kwargs):
+        """Inject parsed metadata to the record."""
+        notes_updated = data.get(self.field, [])
+        notes = deepcopy(record.get(self.field, []))
+        existing_ids = set([note["id"] for note in notes])
+        new_ids = set([note["id"] for note in notes_updated])
+        to_delete = existing_ids - new_ids
+        if sorted(existing_ids) != sorted(new_ids):
```

Comment:
there is a bit of manual checks since the field was designed as a list, each new or removed element need to be checked by its unique id

Reply:
With the discussion we had IRL, the `id` was not needed. Why adding it now?
If not needed, it might be confusing.

Reply:
we had another discussion which we concluded that id is needed to identify the note, the position in the list is not enough (f.e. when one note is deleted and a new added in the same request)

Reply:
What we discussed is that if you mark the deletion sending an empty element, you can rely on the list position:
```
[
    {},   # means delete
    { new one or updated }
]

Reply:
`id` is needed also for identifying which note is new, which already exists, to propagate the timestamp and user on the existing ones (when updating drafts)

Reply:
> `id` is needed also for identifying which note is new, which already exists, to propagate the timestamp and user on the existing ones (when updating drafts)

Not really, as the client will have to send the timestamp and creator, and can be compared with the existing.
It all boils down to what contract you want to have for this API.
The `id` is really necessary only if we need to support re-ordering of the list in conjunction with a insert/update/delete operation.

I am ok with an id, if you can ensure that a previous id cannot be re-used.